src/backend/tty.rs

```diff
@@ -1237,64 +1237,62 @@ impl Tty {
-        let devices = self.devices.get(&node);
-        if let Some(device) = devices {
-            for (_connector, crtc) in device.drm_scanner.crtcs() {
-                let crt_info = device.drm.get_crtc(crtc).unwrap();
-                let gamma_length = crt_info.gamma_length() as usize;
-                let mut red = vec![0; gamma_length];
-                let mut green = vec![0; gamma_length];
-                let mut blue = vec![0; gamma_length];
-                device
-                    .drm
-                    .get_gamma(crtc, &mut red, &mut green, &mut blue)
-                    .unwrap_or_else(|err| warn!("error getting gamma for: {err:?}"));
-                red.extend(green);
-                red.extend(blue);
-                return Some(red);
-            }
+        let tty_state = output.user_data().get::<TtyOutputState>().unwrap();
+        let device = self.devices.get(&tty_state.node)?;
+        let crtc_info = device.drm.get_crtc(tty_state.crtc.clone()).ok()?;
+        let crtc = tty_state.crtc.clone();
+        let gamma_length = crtc_info.gamma_length() as usize;
+        let mut red = vec![0; gamma_length];
+        let mut green = vec![0; gamma_length];
+        let mut blue = vec![0; gamma_length];
+        if let Err(err) = device.drm.get_gamma(crtc, &mut red, &mut green, &mut blue) {
+            warn!("error getting gamma for crtc {crtc:?}: {err:?}");
+            return None;
         }
-        None
+        red.extend(green);
+        red.extend(blue);
+        return Some(red);
     }
 
-    pub fn set_gamma(&self, output: &Output, ramp: Vec<u16>) {
-        let node = output.user_data().get::<TtyOutputState>().unwrap().node;
-
-        let devices = self.devices.get(&node);
-
-        if let Some(device) = devices {
-            for (_connector, crtc) in device.drm_scanner.crtcs() {
-                let crt_info = device.drm.get_crtc(crtc).unwrap();
-                let gamma_length = crt_info.gamma_length() as usize;
-
-                let mut red = ramp.clone();
-                red.truncate(gamma_length);
-                let mut green = ramp.clone();
-                green.drain(0..gamma_length);
-                green.truncate(gamma_length);
-                let mut blue = ramp.clone();
-                blue.drain(0..gamma_length * 2);
-                blue.truncate(gamma_length);
-
-                device
-                    .drm
-                    .set_gamma(crtc, &mut red, &mut green, &mut blue)
-                    .unwrap_or_else(|err| warn!("error setting gamma for: {err:?}"));
-            }
-        }
+    pub fn set_gamma(
+        &self,
+        niri: &mut Niri,
+        output: &Output,
+        ramp: Vec<u16>,
+    ) -> anyhow::Result<()> {
+        let tty_state = output.user_data().get::<TtyOutputState>().unwrap();
+        let device = self
+            .devices
+            .get(&tty_state.node)
+            .ok_or(anyhow!("missing device"))?;
```

Comment:
```suggestion
            .context("missing device")?;
```
I think just this should work?